### ISSUE TITLE
Changes in a WYSIWYG field don't trigger "unsaved changes" warning

### ISSUE THREAD

**Issue by purplespider** [dnadesign/silverstripe-elemental#821]
If you make changes to content in a block's WYSIWYG field, then try and navigate away, e.g. to another page or if you click onto a non-inline block, the alert for unsaved changes doesn't appear and the changes get lost.

Only seems to happen with WYSIWYG fields, other fields trigger the alert fine.

**Comment by purplespider** [comment]
This is still an issue that can lead to editors innocently losing data.

Take the following scenario: an editor is making changes to multiple content blocks on a page. They make changes to a couple of blocks, and then click a third block to make changes, but as this block is not an inline block, it opens in a new screen and all their changes to the other blocks are lost without warning.

**Comment by robertvanlienden** [comment]
Also experiencing this isssue.
For some reason, if I edit a title (textfield) for example, the element gets a "edited" dot AND the save/publish button will change.

If I only edit a WYSIWYG field, only the "edited" dot is showing up, the save/publish buttons will , but for some reason the "saved/published" buttons" not change.

Seems like the "page" is not seeing the change in WYSIWYG editor, but the "dot" indicator on the element is saying me that the element does see the change.
Not sure what is triggering this situation right now...

**Comment by GuySartorelli** [comment]
I haven't been able to reproduce this...
A few questions that might help identify what's happening:

1. What version of silverstripe framework is the site running?
2. What version of elemental are you using?
3. Is it the built-in content block, or some other block?
2. Do you have any modules or configuration that affect the `HTMLEditorField`?
4. Are there any error messages appearing in the developer console of your browser?
5. If you inspect the markup, does the form have the `changed` css class after you make a change to the WYSIWYG content?
6. Are you making the changes to WYSIWYG content directly or using the "source" button to edit the actual markup of the content? (shouldn't matter either way but who knows)
7. Are there any `HTMLEditorField`s on the page other than those in the block?
8. Does this happen when there is only one content block? Or only when there are multiple blocks?

**Comment by purplespider** [comment]
I can still reproduce this on a fresh Silverstripe and Elemental install. e.g.:
`composer create-project silverstripe/installer my-project`
`composer require dnadesign/silverstripe-elemental ^4`

Here's an example:
![Screenshot 2022-02-14 at 11 51 24](https://user-images.githubusercontent.com/329880/153859648-0556d593-4cab-4b25-aba6-e6c934ba49a6.gif)

@GuySartorelli To answer your questions:
> * What version of silverstripe framework is the site running?

4.10.0

> * What version of elemental are you using?

4.8.1

> * Is it the built-in content block, or some other block?

Built-in

> * Do you have any modules or configuration that affect the `HTMLEditorField`?

Nothing else changed/added.

> * Are there any error messages appearing in the developer console of your browser?

Nope, nothing appears in the console.

> * If you inspect the markup, does the form have the `changed` css class after you make a change to the WYSIWYG content?

No. I can see it appearing after changing a non-WYSIWYG, but not on changing a WYSIWYG.

> * Are you making the changes to WYSIWYG content directly or using the "source" button to edit the actual markup of the content? (shouldn't matter either way but who knows)

**Directly:** Issue occurs. ❌
**Via "Source code" button:** Issue doesn't occur, field correctly gets marked as changed. ✅

> * Are there any `HTMLEditorField`s on the page other than those in the block?

Nope. Using default Elemental config which replaces the orig content field.

> * Does this happen when there is only one content block? Or only when there are multiple blocks?

Either. The number of content blocks doesn't seem to affect it. Occurs with 1+. 


Interestingly, if you edit a non-WYSIWYG field immediately after editing a WYSIWYG field it still doesn't mark the page as "changed". So it seems that when editing a WYSIWYG field it _tries_ to do something but this breaks the ability for any field to trigger `changed` until the page is refreshed.

Also note in the example above, as mentioned by @robertvanlienden, when editing a WYSIWYG, the little blue dot to indicate that particular content block _does_ appear, but the form isn't marked as`changed` so maybe something is breaking after setting be blue dot?

**Comment by GuySartorelli** [comment]
Hmm I've tried a few more times and it looks like I _can_ reproduce this, but it doesn't happen every time. It seems to be related to how quickly I type the first few characters - if I just type one letter and wait, the form does get the `changed` class applied and the save/publish buttons update appropriately. But if I spam rapid keypresses, the issue as described occurs.

**Comment by robertvanlienden** [comment]
I can agree with @GuySartorelli.
Today the same behavior happening here.

Needs to get investigated further.

**Comment by purplespider** [comment]
> @purplespider Can you please give it a go just inputting a single character and see if that gives the expected result? I want to make sure it's not just me getting this specific behaviour

Good catch! Yep, I see this too. Entering 1 character and then pausing works fine. But if I enter even just 2 characters in quick succession (i.e. when typing a word!) it breaks.

**Comment by GuySartorelli** [comment]
When the react form has a dirty stage, a hidden `input.element-form-dirty-state` element is created in the inline editable element form.

Our jQuery code has an `onmatch` event handler for `.js-injector-boot .element-editor__container .element-form-dirty-state` (i.e. the hidden input) which fires a change event for the form itself.

This works as expected when adding a single character - the form's change event is caught in `jquery.changetracker.js detectChanges()` and the `change` CSS class is added to the form.

When you type multiple characters in succession, multiple change events are fired for the individual form input fields, a single `input.element-form-dirty-state` is created, the form's change event is fired..... but for some reason I haven't been able to discover yet, it doesn't get caught by `jquery.changetracker.js detectChanges()`.

It's possible this is somehow related to the debouncing that's applied to the `detectChanges` or `handleChanges` functions in changetracker but that's a shot in the dark.

**Comment by emteknetnz** [comment]
Linked PR has been merged it will be automatically tagged shortly
